File: api/src/domain_classes/blueprint.py
from typing import Dict, List

from domain_classes.blueprint_attribute import BlueprintAttribute
from domain_classes.recipe import Recipe, RecipeAttribute
from domain_classes.storage_recipe import DefaultStorageRecipe, StorageRecipe
from enums import StorageDataTypes


def get_storage_recipes(recipes: List[Dict], attributes: List[BlueprintAttribute]):
    if not recipes:
        return [DefaultStorageRecipe(attributes)]
    else:
        return [
            StorageRecipe(
                name=recipe["name"],
                storageAffinity=recipe.get("storageAffinity", StorageDataTypes.DEFAULT.value),
                attributes=recipe["attributes"],
            )
            for recipe in recipes
        ]


def get_ui_recipe(recipes: List[Dict]):
    # TODO: Add a from_dict() on Recipe class. This should not be duplicated.
    return [
        Recipe(
            name=recipe["name"],
            plugin=recipe.get("plugin", "Default"),
            description=recipe.get("description", ""),
            attributes=[
                RecipeAttribute(
                    name=attr["name"],
                    contained=attr.get("contained", True),
                    ui_recipe=attr.get("uiRecipe", None),
                )
                for attr in recipe.get("attributes", [])
            ],
        )
        for recipe in recipes
    ]


class Blueprint:
    def __init__(self, entity: dict):
        self.name = entity["name"]
        self.entity = entity
        self.extends = entity.get("extends", [])
        self.description = entity.get("description", "")
        self.type = entity["type"]
        self.attributes: List[BlueprintAttribute] = [
            BlueprintAttribute.from_dict(attribute) for attribute in entity.get("attributes", [])
        ]
        self.storage_recipes: List[StorageRecipe] = get_storage_recipes(
            entity.get("storageRecipes", []), self.attributes
        )
        self.ui_recipes: List[Recipe] = get_ui_recipe(entity.get("uiRecipes", []))

    def get_non_primitive_types(self) -> List[BlueprintAttribute]:
        blueprints = [attribute for attribute in self.attributes if not attribute.is_primitive]
        return blueprints

File: web/plugins/default-form/src/domain/Blueprint.ts
import { BlueprintAttributeType, BlueprintType, KeyValue } from './types'
import { BlueprintAttribute } from './BlueprintAttribute'

interface IBlueprint {
  getUiAttribute: (name: string, pluginName: string) => object | undefined
}

export class Blueprint implements IBlueprint {
  private attributes: KeyValue = {}
  private uiRecipes: KeyValue = {}
  private blueprintType: BlueprintType

  constructor(blueprintType: BlueprintType) {
    this.blueprintType = blueprintType
    this.addAttributes(this.attributes, blueprintType.attributes)

    if (blueprintType.uiRecipes) {
      blueprintType.uiRecipes.forEach((recipe: any) => {
        const name = recipe.name
        if (name) {
          this.uiRecipes[name] = {}
          if (recipe.attributes) {
            this.addAttributes(this.uiRecipes[name], recipe.attributes)
          }
        }
      })
    }
  }

  public getAttributes() {
    return this.attributes
  }

  private addAttributes(container: KeyValue, attributes: any[]): void {
    if (attributes) {
      attributes.forEach((attr: any) => {
        container[attr.name] = attr
      })
    }
  }

  public getUiAttribute(
    uiRecipeName: string,
    attributeName: string
  ): KeyValue | undefined {
    if (this.uiRecipes && this.uiRecipes[uiRecipeName]) {
      return this.uiRecipes[uiRecipeName][attributeName]
    }
  }

  // helper functions
  isArray(attr: BlueprintAttributeType) {
    return new BlueprintAttribute(attr).isArray()
  }

  public getBlueprintType(): BlueprintType {
    return this.blueprintType
  }

  public getAttribute(name: string): BlueprintAttributeType | undefined {
    return this.attributes[name]
  }

  public getBlueprintAttribute(name: string): BlueprintAttribute | undefined {
    if (this.attributes[name]) {
      return new BlueprintAttribute(this.attributes[name])
    }
    return undefined
  }

  public getUiAttributes(uiRecipeName: string): KeyValue | undefined {
    return this.uiRecipes[uiRecipeName]
  }

  isPrimitive(type: string): boolean {
    //todo use AttributeTypes enum, available in the blueprint.
    return ['string', 'number', 'integer', 'number', 'boolean'].includes(type)
  }

  public validateEntity(entity: KeyValue) {
    this.blueprintType.attributes.forEach(
      (attrType: BlueprintAttributeType) => {
        const attr = new BlueprintAttribute(attrType)
        const value = entity[attr.getName()]
        if (attrType.optional !== true) {
          if (attr.isPrimitive()) {
            if (attrType.default) {
              //todo insert default?  need to do casting.
            }
          } else {
            if (value === undefined && this.isArray(attrType)) {
              entity[attrType.name] = []
            }
          }

          // required
          if (!entity[attrType.name]) {
            console.warn(
              `non optional value is missing for ${attrType.name} of type ${attrType.type}, ${entity.type}`
            )
          }
        }
      }
    )
  }

  /**
   * Index recipe is one to one to blueprints.
   * Not expecting a use case where multiple index recipes for one blueprint is needed.
   * (and complexity is likely to blow up, especially on the api)
   */
  private getIndexRecipeAttributes(): KeyValue | undefined {
    if (this.blueprintType.uiRecipes) {
      const indexRecipe = this.blueprintType.uiRecipes.find(
        (recipe) => recipe.plugin === 'INDEX'
      )
      if (indexRecipe) {
        return this.getUiAttributes(indexRecipe.name)
      }
    }
  }

  /**
   * Filter attributes by ui recipe.
   * The recipes are
   *
   * @param editRecipeName
   */
  public filterAttributesByUiRecipe(editRecipeName = '') {
    const indexAttributes = this.getIndexRecipeAttributes()
    const editAttributes = this.getUiAttributes(editRecipeName)
    return (attr: BlueprintAttribute) => {
      const indexAttribute = indexAttributes && indexAttributes[attr.getName()]
      const editAttribute = editAttributes && editAttributes[attr.getName()]
      /* default edit plugin behavior.
        const INDEX_PRIMITIVE_CONTAINED = false
        const INDEX_ARRAY_CONTAINED = true
        const INDEX_TYPE_CONTAINED = true
      * */
      if (attr.isPrimitive()) {
        return true
      }

      // keep if contained in edit recipe.
      if (editAttribute?.contained) {
        return true
      }

      // return opppsite of index recipe.
      if (indexAttribute?.contained !== undefined) {
        return !indexAttribute.contained
      }
      return false
    }
  }
}
